### PR TITLE
add stratification post

### PR DESCRIPTION
**Pull request by giuseppec**
This post covers three ways to achieve reliable resampling:
  
  1. Stratified resampling for classification problems where each train/test split maintains the target class distribution of the original data set.  
  2. Blocked resampling where a grouping factor determines which observations should be together in train/test splits.  
  3. Using predefined and manually created folds for the train/test splits.

### CODE REVIEW

File: content/post/stratification-blocking.Rmd [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+```{r packages}
+library(mlr3)
+library(mlbench)
+library(tidyverse)
+```
+
+## Stratified resampling
+
+In classification tasks, the ratio of the target class distribution should be similar in each train/test split, which is achieved by stratification.
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]: Maybe a short ref that backs up this statement?

File: content/post/stratification-blocking.Rmd [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+In `mlr3`, each [task](https://mlr3.mlr-org.com/reference/Task.html) has a slot [`col_roles`](https://mlr3.mlr-org.com/reference/Task.html#active-bindings). 
+We can specify the column used for stratification in `task$col_roles$stratum`. 
+This will be illustrated in the following example using the `german_credit` data:  
+
+```{r}
+gc_tsk = tsk("german_credit")
+```
+
+By default, the `col_roles` slot shows the roles of the columns that will be used as `feature`s and the `target` feature:
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]: Maybe more general: col_roles can be more diverse than just feature and target and might even serve multiple roles. So they denote in a general manner the roles that certain variables will have throughout different stages (training is just one).

File: content/post/stratification-blocking.Rmd [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+```{r col_roles}
+gc_tsk$col_roles
+```
+
+We use the target feature called `credit_risk` to specify the column used for stratification:
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]: Can we even use a different one than the target feature for stratification? Would that make sense?

File: content/post/stratification-blocking.Rmd [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+To illustrate if stratification based on multiple categorical features works, we need to instantiate the CV folds again as we changed the features used for stratification:
+
+```{r cv6}
+set.seed(123)
+cv3$instantiate(gc_tsk)
+cv3$instance
+```
+
+Again, we check the relative frequency of observations in each group (combination of `housing` and `telephone`) across each folds:
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]:
```suggestion
Again, we check the relative frequency of observations in each group (combination of `housing` and `telephone`) across all folds:  
```

File: content/post/stratification-blocking.Rmd [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+  left_join(gc_tsk$data() %>% mutate(row_id = 1:n())) %>%
+  group_by(fold) %>%
+  summarise(class_ratio = sum(credit_risk == "bad")/sum(credit_risk == "good"))
+```
+
+And compare it with the target class distribution from the whole data set:
+
+```{r cv3}
+gc_tsk$data() %>%
+  summarise(class_ratio = sum(credit_risk == "bad")/sum(credit_risk == "good"))
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]:
```suggestion
  summarise(class_ratio = sum(credit_risk == "bad") / sum(credit_risk == "good"))
```

File: content/post/stratification-blocking.Rmd [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+gc_tsk$col_roles
+```
+
+We use the target feature called `credit_risk` to specify the column used for stratification:
+
+```{r col_roles_startum}
+gc_tsk$col_roles$stratum = "credit_risk"
+```
+
+After specification of `task$col_roles$stratum` the [`task$strata`](https://mlr3.mlr-org.com/reference/Task.html#active-bindings) active binding will show the number of observations in each group and the corresponding row id's:
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]:
```suggestion
After the specification of `task$col_roles$stratum`, the active binding [`task$strata`](https://mlr3.mlr-org.com/reference/Task.html#active-bindings) will show the number of observations in each group and the corresponding row id's:
```

File: content/post/stratification-blocking.Rmd [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+In the [BreastCancer](https://www.rdocumentation.org/packages/mlbench/versions/2.1-1/topics/BreastCancer) data set, for example, several observations have the same "Id" (Sample code number) which implies these are samples taken from the same patient at different times.
+
+```{r Id}
+# Let's count how many observation actually have the same Id more than once
+sum(table(BreastCancer$Id) > 1)
+```
+
+There are 46 Id's with more than one observation (row).  
+
+The model trained on this data set will be used to predict cancer status in new patients.
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]:
```suggestion
The model trained on this data set will be used to predict cancer status of new patients. 
```

File: content/post/stratification-blocking.Rmd [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+```{r Id}
+# Let's count how many observation actually have the same Id more than once
+sum(table(BreastCancer$Id) > 1)
+```
+
+There are 46 Id's with more than one observation (row).  
+
+The model trained on this data set will be used to predict cancer status in new patients. 
+Hence, we have to make sure that each `Id` occurs exactly in one fold, i.e., all observations with the same `Id` should be either used for training or for evaluating the model.
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]: Maybe try to leave out abbreviations like "i.e." in blog posts.

File: content/post/stratification-blocking.Rmd [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+This way, we get less biased performance estimates via k-fold cross validation.
+This is known as blocked cross validation.
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]: Maybe add that there are different terms for this: I see it often referenced as "Block cross-validation".

"blocked" to me sounds like an adjective blocking something out while "block" more refers to the "blocked" structure of the observations (seen from a clustering perspective). 

Both probably have arguments for their valid existence.

File: content/post/stratification-blocking.Rmd [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+There are 46 Id's with more than one observation (row).  
+
+The model trained on this data set will be used to predict cancer status in new patients. 
+Hence, we have to make sure that each `Id` occurs exactly in one fold, i.e., all observations with the same `Id` should be either used for training or for evaluating the model.
+This way, we get less biased performance estimates via k-fold cross validation.
+This is known as blocked cross validation. 
+The following example will illustrate blocked cross validation which can be achieved by specifying a blocking factor in the [`task$col_roles$group`](https://mlr3.mlr-org.com/reference/Task.html#active-bindings) slot:  
+
+```{r create task_ blocked}
+# Use Id column as grouping
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]: Careful with the term "grouping". This is used in mlr2 as a dedicated, different resampling strategy.

File: content/post/stratification-blocking.Rmd [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+bc_tsk$data() %>%
+  merge(cv5$instance, by.x = "Id", by.y = "row_id") %>%
+  group_by(Id) %>%
+  summarize(unique_folds = length(unique(fold))) %>%
+  filter(unique_folds > 1)
+```
+
+As expected, the table is empty as there are no Id's present in more than one fold.
+
+## Resampling with predefined folds
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd:186]: This is what we called "grouping" in mlr2 - but the name is not really backed up by any literature.

File: content/post/stratification-blocking.Rmd [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+```{r predefined_folds1}
+set.seed(1)
+folds = sample(rep(1:5, length.out = nrow(BreastCancer)),
+  size = nrow(BreastCancer),
+  replace = F)
+head(folds, 20)
+table(folds)
+```
+
+This vector is now added to the data set and will be used as blocking factor:
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]: Careful with using "blocking" here again - possible confusion.

File: content/post/stratification-blocking.Rmd (hunk 11 above)
Comment (pat-s) [inline content/post/stratification-blocking.Rmd:186]: This is what we called "grouping" in mlr2 - but the name is not really backed up by any literature.

File: content/post/stratification-blocking.Rmd (hunk 10 above)
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]: Careful with the term "grouping". This is used in mlr2 as a dedicated, different resampling strategy.

File: content/post/stratification-blocking.Rmd [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+```{r predefined_folds3}
+bc_tsk$data() %>%
+  merge(cv5$instance, by.x = "foldIds", by.y = "row_id") %>%
+  group_by(foldIds) %>%
+  summarize(unique_folds = length(unique(fold)))
+```
+
+There are five groups and each `foldIds` appears only in exactly one fold, i.e., each instantiated fold corresponds to one of the predefined folds.
+
+The previous example does not cover how to perform repeated k-fold CV or time series CV with predefined indices.
```
Comment (pat-s) [inline content/post/stratification-blocking.Rmd:234]: Also note that no repetitions are possible with this method.

File: content/post/stratification-blocking.Rmd (hunk 10 above)
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]: Careful with the term "grouping". This is used in mlr2 as a dedicated, different resampling strategy.

File: content/post/stratification-blocking.Rmd (hunk 12 above)
Comment (pat-s) [inline content/post/stratification-blocking.Rmd]: Careful with using "blocking" here again - possible confusion.

File: content/post/stratification-blocking.Rmd (hunk 13 above)
Comment (pat-s) [inline content/post/stratification-blocking.Rmd:234]: Also note that no repetitions are possible with this method.

**Review by pat-s — COMMENTED** [review]
Thanks! Good read. 

Thanks also for sticking to the format specifications - it's rare that I do not have to comment on this 🙈

File: content/post/stratification-blocking.Rmd (hunk 3 above)
Comment (giuseppec) [reply content/post/stratification-blocking.Rmd]: We do this with housing and telephon in the following example

File: content/post/stratification-blocking.Rmd (hunk 1 above)
Comment (giuseppec) [reply content/post/stratification-blocking.Rmd]: done

File: content/post/stratification-blocking.Rmd (hunk 2 above)
Comment (giuseppec) [reply content/post/stratification-blocking.Rmd]: we included this comment, thx.

File: content/post/stratification-blocking.Rmd (hunk 4 above)
Comment (giuseppec) [reply content/post/stratification-blocking.Rmd]: done

File: content/post/stratification-blocking.Rmd (hunk 5 above)
Comment (giuseppec) [reply content/post/stratification-blocking.Rmd]: fixed

File: content/post/stratification-blocking.Rmd (hunk 6 above)
Comment (giuseppec) [reply content/post/stratification-blocking.Rmd]: fixed

File: content/post/stratification-blocking.Rmd (hunk 7 above)
Comment (giuseppec) [reply content/post/stratification-blocking.Rmd]: fixed

File: content/post/stratification-blocking.Rmd (hunk 9 above)
Comment (giuseppec) [reply content/post/stratification-blocking.Rmd]: done.

File: content/post/stratification-blocking.Rmd (hunk 10 above)
Comment (giuseppec) [reply content/post/stratification-blocking.Rmd]: done

**Review by pat-s — APPROVED** [review]
Ready from your side @giuseppec @missuse?

**Comment by missuse** [comment]
> Ready from your side @giuseppec @missuse?

I don't see any problems. Although, I must admit, I usually see many problems only after I click submit/merge/publish

**Pull request merged**